flink-runtime/src/main/java/org/apache/flink/runtime/leaderretrieval/DefaultLeaderRetrievalService.java

```diff
@@ -0,0 +1,155 @@
+
+	/**
+	 * Creates a default leader retrieval service with specified {@link LeaderRetrievalDriver}.
+	 *
+	 * @param leaderRetrievalDriver {@link LeaderRetrievalDriver} implemented by Zookeeper, Kubernetes, etc.
+	 */
+	public DefaultLeaderRetrievalService(LeaderRetrievalDriver leaderRetrievalDriver) {
+		this.leaderRetrievalDriver = leaderRetrievalDriver;
+
+		this.lastLeaderAddress = null;
+		this.lastLeaderSessionID = null;
+
+		running = false;
+
+		this.leaderRetrievalStateHandler = new StateHandler();
+	}
+
+	@Override
+	public void start(LeaderRetrievalListener listener) throws Exception {
+		Preconditions.checkNotNull(listener, "Listener must not be null.");
+		Preconditions.checkState(leaderListener == null, "DefaultLeaderRetrievalService can " +
+			"only be started once.");
+
+		LOG.info("Starting DefaultLeaderRetrievalService with {}.", leaderRetrievalDriver);
+
+		synchronized (lock) {
+			leaderListener = listener;
+			leaderRetrievalDriver.start(leaderRetrievalStateHandler);
+
+			running = true;
+		}
+	}
+
+	@Override
+	public void stop() throws Exception {
+		synchronized (lock) {
+			if (!running) {
+				return;
+			}
+			running = false;
+
+			LOG.info("Stopping DefaultLeaderRetrievalService {} with {}.", leaderRetrievalDriver);
+
+			leaderRetrievalDriver.stop();
+		}
+	}
+
+	/**
+	 * Helper class for the specific {@link LeaderRetrievalDriver} to notify the leader changes.
+	 */
+	public class StateHandler {
+
+		/**
+		 * Called by specific {@link LeaderRetrievalDriver} to notify new leader address.
+		 * @param supplier supplier to provide a {@link Tuple2} of leader information. f0 is UUID, and f1 is leader
+		 * address. The exception will be handled by leader listener.
+		 */
+		@GuardedBy("lock")
+		public void notifyIfNewLeaderAddress(SupplierWithException<Tuple2<UUID, String>, Exception> supplier) {
+			try {
+				final Tuple2<UUID, String> leaderInformation = supplier.get();
+				final UUID newLeaderSessionID = leaderInformation.f0;
+				final String newLeaderAddress = leaderInformation.f1;
+				if (!(Objects.equals(newLeaderAddress, lastLeaderAddress) &&
+					Objects.equals(newLeaderSessionID, lastLeaderSessionID))) {
+					if (newLeaderAddress == null && newLeaderSessionID == null) {
+						LOG.debug("Leader information was lost: The listener will be notified accordingly.");
+					} else {
+						LOG.debug(
+							"New leader information: Leader={}, session ID={}.",
+							newLeaderAddress,
+							newLeaderSessionID);
+					}
+
+					lastLeaderAddress = newLeaderAddress;
+					lastLeaderSessionID = newLeaderSessionID;
+					leaderListener.notifyLeaderAddress(newLeaderAddress, newLeaderSessionID);
+				}
+			} catch (Exception e) {
+				leaderListener.handleError(e);
+			}
+		}
+
+		/**
+		 * Called by specific {@link LeaderRetrievalDriver} to notify leader loss.
+		 */
+		@GuardedBy("lock")
+		public void notifyLeaderLoss() {
+			notifyIfNewLeaderAddress(() -> new Tuple2<>(null, null));
+		}
```

Comment:
Isn't this call not redundant? Couldn't we use the `notifyLeaderAddress(LeaderAddressInformation)` which we call with `notifyLeaderAddress(LeaderAddressInformation.empty())` or so?